contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+These plugins do not return any value and are not expected to influence the
+scheduling decision made in the scheduling cycle.
+
+# USE-CASES
+
+In this section we provide a couple of examples on how the scheduling framework
+can be used to solve common scheduling scenarios.
+
+### Dynamic binding of cluster-level resources
+
+Cluster level resources are resources which are not immediately available on
+nodes at the time of scheduling Pods. Scheduler needs to ensure that such
+cluster level resources are bound to a chosen Node before it can schedule a Pod
+that requires such resources to the Node. We refer to this type of binding of
+resources to Nodes at the time of scheduling Pods as dynamic resource binding.  
+Dynamic resource binding has proven to be a challenge in Scheduler v1, because
+Scheduler v1 is not flexible enough to support various types of plugins at
+different phases of scheduling. As a result, binding of storage volumes is
+integrated in the scheduler code and some non-trivial changes are done to the
+scheduler extender to support dynamic binding of network GPUs.  
+The scheduling framework allows such dynamic bindings in a cleaner way. The main
+thread of scheduling framework process a pending Pod that requests a network
+resource and finds a node for the Pod and reserves the Pod. A dynamic resource
+binder plugin installed at "Pre-Bind" stage is invoked (in a separate thread).
+It analyzes the Pod and when detects that the Pod needs dynamic binding of the
+resource, the plugin tries to attach the cluster resource to the chosen node and
+then returns true so that the Pod can be bound. If the resource attachment
+fails, it returns false and the Pod will be retried.  
+When there are multiple of such network resources, each one of them installs one
+"pre-bind" plugin. Each plugin looks at the Pod and if the Pod is not requesting
+the resource that they are interested in, they simply return "true" for the
+pod.
+
+### Gang Scheduling
+
+Gang scheduling allows a certain number of Pods to be scheduled simultaneously.
+If all the members of the gang cannot be scheduled at the same time, none of
+them should be scheduled. Gang scheduling may have various other features as
+well, but in this context we are interested in simultaneous scheduling of Pods.  
+Gang scheduling in the scheduling framework can be done with an "Admit" plugin.
+The main scheduling thread processes pods one by one and reserves nodes for
+them. The gang scheduling plugin at the admit stage is invoked for each pod.
+When it finds that the pod belongs to a gang, it checks the properties of the
+gang. If there are not enough members of the gang which are scheduled or in
+"wait" state, the plugin returns "wait". When the number reaches the desired
+value, all the Pods in wait state are approved and sent for binding.
+
+# OUT OF PROCESS PLUGINS
+
+Out of process plugins (OOPP) are called via JSON over an HTTP interface. In
```

Comment:
Why JSON?

Reply:
I am open to explore ideas here and don't have a strong opinion about one or the other. That said, there are a several reasons I think JSON might be a better choice here:

1. One cannot expect very high performance from OOPPs. If high performance is required, we highly recommend in-process plugins. If performance is not an issue, JSON should work fine.
1. Protobuf requires the addition of all the protobuf (and probably the rest of GRPC) libraries.
1. Webhooks are more aligned with other K8s extensions.